src/metabase/lib/portable.cljc

```diff
@@ -0,0 +1,124 @@
+(ns metabase.lib.portable
+  (:require
+   [medley.core :as m]
+   [metabase.lib.metadata :as lib.metadata]
+   [metabase.lib.schema :as lib.schema]
+   [metabase.lib.util.match :as lib.util.match]
+   [metabase.lib.walk :as lib.walk]
+   [metabase.util.malli :as mu]
+   [metabase.util.malli.registry :as mr]))
+
+(defprotocol ExportVisitor
+  (export-database     [this mp context database-id])
+  (export-source-table [this mp context source-table-id])
+  (export-source-card  [this mp context source-card-id])
+  (export-field        [this mp context field-id]))
+
+(defn export-visitor?
+  "Does x implement the ExportVisitor protocol?"
+  [x]
+  #?(:clj  (extends? ExportVisitor (class x))
+     :cljs (satisfies? ExportVisitor x)))
+
+(mr/def ::export-visitor
+  "Schema for things that satisfy the [[metabase.lib.portable/ExportVisitor]] protocol."
+  [:fn
+   {:error/message "Valid ExportVisitor"}
+   #'export-visitor?])
+
+(defprotocol ImportVisitor
+  (import-database     [this mp context database-info])
+  (import-source-table [this mp context source-table-info])
+  (import-source-card  [this mp context source-card-info])
+  (import-field        [this mp context field-info]))
+
+(defn import-visitor?
+  "Does x implement the ImportVisitor protocol?"
+  [x]
+  #?(:clj  (extends? ImportVisitor (class x))
+     :cljs (satisfies? ImportVisitor x)))
+
+(mr/def ::import-visitor
+  "Schema for things that satisfy the [[metabase.lib.portable/ImportVisitor]] protocol."
+  [:fn
+   {:error/message "Valid ImportVisitor"}
+   #'import-visitor?])
+
+(defn- export-walk-database [query context export-visitor]
+  (let [mp (lib.metadata/->metadata-provider query)]
+    (m/assoc-some query :database (some->> (:database query)
+                                           (export-database export-visitor mp context)))))
```

Comment:
I was about to say "we have a util for this" but we kind of don't. There's `m/update-existing` and `u/assoc-dissoc`.

Depending on the intent here, `m/update-existing` might be neater?
`(m/update-existing query :database #(export-database export-visitor mp context %))`

The difference from what you have is that an explicit `:database nil` would break it. I don't think that ever actually happens?

At any rate, this form appears a lot, so let's make it a local helper at least.